### ISSUE TITLE
Create new temporary directory/resource extension

### ISSUE THREAD

**Issue by nipafx** [junit-pioneer/junit-pioneer#348]
After Marc Philipp implemented the `@TempDir` extension in Pioneer, it was later also added to Jupiter. It didn't (and as of September 2020 doesn't) support custom file systems, but other than that is more feature-rich than Pioneer's variant. This creates an unpleasant situation where there are two extensions that solve the same problem with the same approach, but slightly different feature sets (and an identical name to boot).

That said, there are a lot of feature requests for Jupiter's variant (⚠️ _some of these are already implemented!_ ⚠️):

* [Multiple declarations of `@TempDir` in same parameter list or class declaration should fail](https://github.com/junit-team/junit5/issues/2219)
* [Support @TempDir on constructor parameters](https://github.com/junit-team/junit5/issues/1754)
* [Introduce cleanup mode in @TempDir](https://github.com/junit-team/junit5/issues/1814)
* [Multiple @TempDirs should result in multiple directories](https://github.com/junit-team/junit5/issues/1967)
* [Provide support for setting a custom parent directory for @TempDir](https://github.com/junit-team/junit5/issues/2088)
* [Support disabling cleanup for @TempDir](https://github.com/junit-team/junit5/issues/2159)
* [Support custom file system for `@TempDir`](https://github.com/junit-team/junit5/issues/2400)
* [Make `@TempDir` restore mint permissions before closing context to avoid `AccessDeniedException`](https://github.com/junit-team/junit5/issues/2609)

After evaluating the situation in #277, we decided to remove the extension before the 1.0 release with the goal to look for a better solution to the problem in the future (i.e. this issue). So the task is to come up with a temporary directory (or, more generally temporary resource) extension that covers _more_ than what Jupiter does, possibly with an approach that makes different trade-offs (so users of Jupiter and Pioneer have more choices to solve their specific problem).

One interesting proof of concept to look at is @sormuras': [1](https://github.com/sormuras/brahms#temporary-directory-as-resource), [2](https://github.com/sormuras/brahms/blob/master/src/test/java/integration/resource/ChickenAndEggTests.java)

**Comment by nipafx** [comment]
For starters, we're just looking for a design proposal or proof of concept that shows the outline of the extension and how it could provide the features requested above.

**Comment by jbduncan** [comment]
I just wanted to say that this issue is of interest to me, and that I've just signed up for hack.commit.push so that I can dedicate some time to work on it.

I'll see if I can find the time beforehand to submit a proposal, but worst case scenario I'll draft it up on the day.

Is this the right place for me to discuss proposals, or would Discord or some place else be better? :)

**Comment by Michael1993** [comment]
Feel free to discuss it here, on Discord, on Twitter, via smoke signals, over telegram or carrier pigeon (or any other way). We'd love to hear from you!

Personally, I'd use Discord, for quick/casual feedback and the GitHub issue for a more formal proposal (because the latter is a bit more permanent) but any way you want is fine.

**Comment by nipafx** [comment]
> but any way you want is fine.

I want to amend this: In the end, a summary of the conversation and the proposal needs to land here for future reference. So while preliminary conversations are a better fit for Discord et al, showing the actual proposal here would be easiest.

**Comment by keturn** [comment]
I had a want for this the other day: https://github.com/junit-team/junit5/issues/1786 : Support @TempDir tags in extensions

**Comment by jbduncan** [comment]
Just wanted to follow up and say that I'm still interested in this, I've just been feeling tired from work lately, so I've dedicated some time in my calendar to revisit this a week before hack.commit.push to get an initial proposal out there.

**Comment by jbduncan** [comment]
When I do get around to this and submit my proposal, I should ensure that my implementation satisfies the following JUnit 5 bug report too: https://github.com/junit-team/junit5/issues/2609.

@nipafx Would you mind amending the OP to link to this issue as well? :)

**Comment by nipafx** [comment]
@jbduncan Done.

**Comment by jbduncan** [comment]
I now have an initial proposal for what an API for solving this overall issue could look like. (See below.)

Many thanks to @sormuras for the inspiration with his [proof-of-concept](https://github.com/sormuras/brahms#temporary-directory-as-resource)!

(I've not yet taken all the issues linked in the OP into account, so any feedback on anything that I've missed would be most appreciated!

I'm also open to feedback on anything that's unclear or too complicated.)

[issue-348-proposal.md](https://github.com/junit-pioneer/junit-pioneer/files/6528763/issue-348-proposal.md)

**Comment by jbduncan** [comment]
My most up-to-date work can be found at https://github.com/jbduncan/junit-pioneer/tree/issue/348-temp-dir-extension.

**Comment by sormuras** [comment]
For the record: using my own dog-food "internally" here https://github.com/sormuras/bach/tree/main/test.base/test/java/test/base/resource

**Comment by jbduncan** [comment]
...and I've just seen I forgot to explain something called `@Singleton` in my proposal, so I'll see if I can submit an amendment later this week.

**Comment by jbduncan** [comment]
Here's the new version. [issue-348-proposal.md](https://github.com/junit-pioneer/junit-pioneer/files/6542314/issue-348-proposal.md)

Or if you prefer, here's the associated [Git diff](https://github.com/jbduncan/junit-pioneer/commit/7501c015b300b3d051b329083f657ae21b5cb83c) and the above new version as a [branch on Git](https://github.com/junit-pioneer/junit-pioneer/compare/main...jbduncan:issue/348-temp-dir-extension).

**Comment by jbduncan** [comment]
I've decided to not include the before-mentioned `@Singleton` annotation, because I'm not sure if I'll have time to implement it on top of the rest of the proposal.

**Comment by jbduncan** [comment]
In response to feedback during hack.commit.push, here's my latest proposal!

Git diff: https://github.com/jbduncan/junit-pioneer/commit/22ed80ce7b4e0e757124e50b2d03b41b95109677
Entire Git branch: https://github.com/junit-pioneer/junit-pioneer/compare/main...jbduncan:issue/348-temp-dir-extension

**Comment by nipafx** [comment]
I'm just gonna publish your proposal here. Verbatim in the next message.

**Comment by nipafx** [comment]
Extension `ResourceManager` looks for fields and
parameters annotated with `@New` `@Share` and `@Shared`.

`@New` is a field/parameter-level annotation that accepts
a Class that implements `ResourceSupplier<T>`. For example,
`@New(TemporaryDirectory.class)`

`@Share` is a class-level annotation that accepts a Class
that implements `ResourceSupplier<T>`, as well as a String
name. For example,
`@Share(name = "aClassWideTempDirectory", value = TemporaryDirectory.class)`

`@Shared` is a field/paramter-level annotation that accepts
a String value. For example,
`@Shared("aClassWideTempDirectory")`

When a parameter in a test method is annotated with `@New`,
it checks if the given class implements `ResourceSupplier`
and has a no-args constructor. If so, it creates a new
instance of the `ResourceSupplier` implementation and asks
it for a `T` with `ResourceSupplier::get`. If `@New` is
used on a test method parameter, the associated
`ResourceSupplier` (and thus the contained `T`) only lasts
as long as the test does and is "torn down" straight
afterwards. If `@New` is used on a test class field, the
associated `ResourceSupplier` is instantiated anew before
and "torn down" immediately after each and every test
method in the test class.

When a test class is annotated with `@Share`, it checks if
the given class implements `ResourceSupplier` and has a
no-args constructor. If so, the `ResourceSupplier` instance
will be instantiated and saved until all the test methods
in that class have been run, at which point it will be
"torn down". The `ResourceSupplier` (and thus the contained
`T`) is saved with a key, where this key comes from
`@Share`'s String name field. This allows the `T` to be
reused across tests in the test class.

When a field or test method parameter is annotated with
`@Shared`, it finds a `@Share` annotation on the same test
class whose String name matches the `@Shared`'s String
value. If such a `@Share` can be found, then the field or 
parameter annotated with `@Shared` is populated with the 
`T` from the `ResourceSupplier` from the `@Share`.

When a `@New` or `@Share` is "torn down",
it calls the `ResourceSupplier` implementation's `close`
method on the field/parameter's instance of `T`.

For example:

```java
@ExtendWith(ResourceManager.class)
class FooTests {

  // Before each test method, this annotation creates a new
  // instance of `Path` that points to a new subdirectory of
  // the machine's temporary directory.
  //
  // After each test method, it is torn down, ready to be recreated
  // for the next test method.
  @New(TemporaryDirectory.class)
  Path firstDirectory;
  
  @Test
  void testFoo1(
      // Creates a `Path` pointing to another
      // temporary directory.
      //
      // It is created before this test method starts,
      // and is closed as soon as this method is finished.
      @New(TemporaryDirectory.class) Path secondDirectory,

      // @Dir is a shortcut for @New(TempDirectory.class).
      @Dir Path thirdDirectory,

      // Thus fourthDirectory is different to thirdDirectory.
      @Dir Path fourthDirectory, 
      
      // Creates a new resource for the duration of this
      // test method, which is provided by a new, user-defined
      // `InMemoryDirectory` resource supplier.
      // (See InMemoryDirectory class below.)
      @New(InMemoryDirectory.class) Path inMemoryDirectory) {
    // ...
  }
}

@ExtendWith(ResourceManager.class)
@Share(name = "aClassWideTempDirectory", value = TemporaryDirectory.class)
class BarTests {
  @Test
  void testBar(
      // The key "aClassWideTempDirectory" below references
      // the @Share annotation above.
      // Thus the temporary directory here...
      @Shared("aClassWideTempDirectory")
      Path aClassWideTempDirectory) {
    Files.writeString(aClassWideTempDirectory.resolve("bar.txt"), "bar1");
  }
  
  @Test
  void testBar2(
      // ...is the same temporary directory as here!
      @Shared("aClassWideTempDirectory")
      Path aClassWideTempDirectory) {
    Files.writeString(aClassWideTempDirectory.resolve("bar.txt"), "bar2");
  }
}

// At the end, "aClassWideTempDirectory" will have a
// "bar.txt" file with lines "bar1" and "bar2".
```

```java
// This ResourceSupplier comes out of the box with this extension.
public final class TemporaryDirectory implements ResourceSupplier<Path> {
  
  private final Path path;
  
  public TemporaryDirectory() {
    // creates a new subdirectory on the machine-wide
    // temporary directory
    this.path = ...
  }
  
  @Override
  public Path get() {
    // returns the subdirectory above
    return path;
  }
  
  @Override
  public void close() {
    // deletes the subdirectory
    deleteRecursively(path);
  }
  
  private static void deleteRecursively(Path path) {
    // ...
  }
}
```

```java
import com.google.common.jimfs.Configuration;
import com.google.common.jimfs.Jimfs;

// An example of a user-defined resource supplier that uses Google's
// jimfs in-memory filesystem. Used in FooTests above.
public final class InMemoryDirectory implements ResourceSupplier<Path> {
  
  // An example of a "resource" that must eventually be closed
  // to prevent resource starvation.
  private final FileSystem jimFs;
  
  public InMemoryDirectory() {
    this.jimFs = JimFs.newFileSystem(Configuration.unix());
  }
  
  @Override
  public Path get() {
    return jim.getPath("/");
  }
  
  @Override
  public void close() {
    try {
      jim.close();
    } catch (Exception e) {
      throw new RuntimeException("Cannot close in-memory filesystem", e);
    }
  }
  
}
```

```java
import mockwebserver3.MockWebServer;

// We can even create a resource supplier that holds an OkHttp mock web server!
public final class WebServer implements ResourceSupplier<Path> {
    
  // Another example of a resource that eventually needs to be closed.
  private final MockWebServer mockWebServer;
  
  public WebServer() {
    this.mockWebServer = new MockWebServer();
    this.mockWebServer.start();
  }
  
  @Override
  public MockWebServer get() {
    return mockWebServer;
  }
  
  @Override
  public void close() {
    mockWebServer.close();
  }
  
}

// Example test
@ExtendWith(ResourceManager.class)
class BazTests {
  @Test
  void testBaz(
      @New(WebServer.class) MockWebServer mockWebServer) {
    // given
    mockWebServer.setDispatcher(
        new Dispatcher() {
          @Override
          public MockResponse dispatch(RecordedRequest request) throws InterruptedException {
            return new MockResponse().setResponseCode(404);
          }
        });
    
    // when
    var client = new MyOwnHttpClient(mockWebServer.url("/"));
    MyOwnHttpClientResponse response = client.get();
    
    // then
    assertTrue(response.is404());
  }
}
```

```java
// The basic building block of this extension.
public interface ResourceSupplier<T> {

  /**
   * Returns an instance of <code>T</code> pointing to a resource managed by
   * <code>this</code> instance of <code>ResourceSupplier</code>.
   * 
   * @return an instance of <code>T</code>.
   */
  T get();

  /**
   * Closes the underlying resource associated with the <code>T</code>
   * returned by {@link #get()}.
   */
  void close();
}
```

**Comment by nipafx** [comment]
I really like this proposal. I do recommend a different approach to closing resources, though:

```java
interface Resource<T> extends Supplier<T>, AutoCloseable {

	@Override
	default void close() throws Exception {
		// no op by default
	}

}

interface ResourceFactory<T> extends AutoCloseable {

	Resource<T> create() throws Exception;

	@Override
	default void close() throws Exception {
		// no op by default
	}

}
```

Then, in the `ParameterProvider`:

```java
@Override
public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
	// we get this from, e.g. `@Named(PathSupplier.class)`
	Class<ResourceFactory<?>> supplierType = null;
	ResourceFactory<?> factory = supplierType.getConstructor().newInstance();
	extensionContext.getStore(null).put("the factory", factory);

	Resource<?> resource = factory.create();

	// say that resource needs cleanup
	// 	~> generally, we can't assume the resource itself is `AutoCloseable`
	//  ~> adding `p` to the store will not lead to cleanup
	extensionContext.getStore(null).put("the resource", resource);

	return resource.get();
}
```

Btw, adding shared resources to the store will make it easy to retrieve them from there.

**Comment by nipafx** [comment]
Another thought - how can we have arguments for resource creation, e.g.:

```java
@Test
// how do we specify the name of the file?
void testFoo(@New(TemporaryFile.class) Path file) {
	// ...
}
```

Oh wait, can we just have an attribute `Object[] arguments` on `@New` whose values get passed to the `ResourceFactory` constructor?

**Comment by jbduncan** [comment]
For posterity, I discussed @nipafx's proposal above with him over his Twitch stream, and I'm happy to give it a shot!

**Comment by jbduncan** [comment]
I've just created a draft PR with my ongoing work: https://github.com/junit-pioneer/junit-pioneer/pull/491

Even though it's a draft, feedback is very much welcome.

**Comment by jbduncan** [comment]
Real life commitments are getting in the way of me progressing with this issue right now, so I'll let you know when I've found the time to work on this again.

**Comment by jbduncan** [comment]
As a future note for myself, it looks like JUnit 5 has a [`CloseableResource`](https://github.com/junit-team/junit5/blob/43638eb6a870e0d6c49224053dfeb39dcf0ef33f/junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ExtensionContext.java#L405) type that I could reuse as part of my work over at https://github.com/junit-pioneer/junit-pioneer/pull/491. It looks safe to use as it's marked as stable.

**Comment by jbduncan** [comment]
Note for myself: I thought at some point that `@Share` and `@Shared` were redundant, because I can adapt the following example:

```java

@ExtendWith(ResourceManager.class)
@Share(name = "aClassWideTempDirectory", value = TemporaryDirectory.class)
class BarTests {
  @Test
  void testBar(
      @Shared("aClassWideTempDirectory")
      Path aClassWideTempDirectory) {
    Files.writeString(aClassWideTempDirectory.resolve("bar.txt"), "bar1");
  }
  
  @Test
  void testBar2(
      @Shared("aClassWideTempDirectory")
      Path aClassWideTempDirectory) {
    Files.writeString(aClassWideTempDirectory.resolve("bar.txt"), "bar2");
  }
}
```

...to this, using `@New` and a test class constructor:

```java

@ExtendWith(ResourceManager.class)
class BarTests {

  BarTests(@New(TemporaryDirectory.class) Path aClassWideTempDirectory) {
    
  }

  @Test
  void testBar() {
    Files.writeString(aClassWideTempDirectory.resolve("bar.txt"), "bar1");
  }
  
  @Test
  void testBar2() {
    Files.writeString(aClassWideTempDirectory.resolve("bar.txt"), "bar2");
  }
}
```

However, this doesn't help if a resource needs to be used across more than one test class, like a database. So I'll pursue `@Share(d)` after I'm done with `@New`, maybe in a subsequent PR to break things up a bit.

**Comment by jbduncan** [comment]
Here is my TODO list for my [PR](https://github.com/junit-pioneer/junit-pioneer/pull/491):

- [ ] I currently use `System::identityHashCode` to uniquely identify resources and resource factories, but this is [error-prone](https://github.com/plume-lib/plume-util/blob/ccdb37b4393bdbbc0087e3317b9a1d1b4827b78b/src/main/java/org/plumelib/util/UniqueIdMap.java) apparently, so use an atomic counter like `AtomicLong` instead.

- [ ] Consider searching junit-pioneer for similar usages of `(identity)HashCode` and `Atomic(Integer|Long)`, refactoring those usages out into a `UniqueIdMap` class, and raising it as a suggestion in a new PR.

- [ ] Finish all TODOs in my PR.

- [ ] Ensure that my PR tackles these issues:
  - [ ] https://github.com/junit-team/junit5/issues/2219
  - [ ] https://github.com/junit-team/junit5/issues/1754
  - [ ] https://github.com/junit-team/junit5/issues/1814
  - [ ] https://github.com/junit-team/junit5/issues/1967
  - [ ] https://github.com/junit-team/junit5/issues/2088
  - [ ] https://github.com/junit-team/junit5/issues/2159
  - [ ] https://github.com/junit-team/junit5/issues/2400
  - [ ] https://github.com/junit-team/junit5/issues/2609

**Comment by nipafx** [comment]
While [reviewing this](https://github.com/junit-pioneer/junit-pioneer/pull/491#pullrequestreview-716186194), I wondered whether we also need a way to prevent users of shared resources to run tests that use them in parallel. The only way I could come up with to do that would be to add a lock to `Resource` that we can then use to synchronize on. Use of the lock could either be activated by based on the resource (e.g. files need synchronizations, directories don't) or by the user (e.g. `@Shared` could get an attribute `boolean lock()`).

**Comment by jbduncan** [comment]
@nipafx Ah, I hadn't considered parallel execution yet! Thanks for starting this conversation. It's in the ever-growing [TODO list](https://github.com/junit-pioneer/junit-pioneer/issues/348#issuecomment-884922323) now, for me to think about or discuss later.

**Comment by jbduncan** [comment]
Great news, I've fixed [the illusive bug](https://github.com/junit-pioneer/junit-pioneer/issues/348#issuecomment-938925966)! So now the only thing I think I have to do is address the thread safety.

**Comment by jbduncan** [comment]
I've been on @nipafx's Twitch stream tonight - thank you very much for your help, Nicolai! - but I now have to solve some other seemingly concurrency-related bugs that have cropped up (noted in my uber TODO list further up). I'll report back when I have the time to delve into things further.
